### PR TITLE
Testing libp2p-spdy on the browser

### PR DESCRIPTION
**Pull request by xicombd**
Re-opening #5, from a branch inside this repo.

#### [Quoting @diasdavid](https://github.com/diasdavid/js-libp2p-spdy/pull/5#issuecomment-199367903):

> We've investigated and currently something weird happens where spdy streams inside the browser work if they are handshaked over stream-pairs, however when connecting a browser - node.js over websockets, opening a stream from the browser succeeds while opening a stream from Node.js fails. This could indicate a problem with the underlying transport, however, our tests with https://github.com/diasdavid/js-libp2p-multiplex and libp2p-websockets work flawlessley.

- [x] browser - browser
- [x] browser - node (fails on node opening a stream)

### CODE REVIEW

File: tests/spdy-over-ws-test.js [hunk 1]
```diff
@@ -62,6 +67,43 @@ describe('spdy', () => {
...
+  })
+
+  it('open a multiplex stream from listener and write to it', (done) => {
+    dialer.once('stream', (conn) => {
+      conn.pipe(conn)
+    })
+
+    const conn = listener.newStream()
+    conn.write('hey')
+    conn.write('ho')
```
Comment (diasdavid) [inline tests/spdy-over-ws-test.js]: @xicombd this is not enough, they just get buffered and written on the same packet. It needs to be more stress tested

**Comment by diasdavid** [comment]
Ok, state of things:

We have a bug when opening a stream from the Server to the Browser in the spdy multiplex connection established through libp2p-websockets:

```sh
Uncaught Error: need dictionary:  (/Users/david/code/protocol-labs/_libp2p/js/js-libp2p-spdy/tests/browser-nodejs/browser.js:3061)
Error: Uncaught Error: need dictionary:  (/Users/david/code/protocol-labs/_libp2p/js/js-libp2p-spdy/tests/browser-nodejs/browser.js:3061)
```

So far, we've ruled out

- [x] libp2p-websockets is not broken. We have run several tests on it and we've used libp2p-multiplex, another stream muxer, to run the same tests that we are doing with spdy and everything works like a charm
- [x] libp2p-spdy works in the browser. We've run tests running spdy on top of a stream-pair inside the browser and everything goes fine.
- [x] libp2p-spdy works on top of libp2p-websockets, if ran on Node.js to Node.js

The only thing that causes that error is really opening a libp2p-websockets connection from browser to a node.js process, and then opening a multiplexed stream from node.js. Very important to note that if we open a multiplex spdy stream from the browser to node.js, it works well.

HALP 😱

**Comment by dignifiedquire** [comment]
Started debugging the issue. My first finding is that the browser receives a message that is not inflatable by `zlib.inflate`. The message is

```js
var buf = new Buffer([56, 48, 227, 198, 167, 194, 0, 103, 0, 152, 255, 0, 0, 0, 5, 0, 0, 0, 7, 58, 109, 101, 116, 104, 111, 100, 0, 0, 0, 4, 80, 79, 83, 84, 0, 0, 0, 8, 58, 118, 101, 114, 115, 105, 111, 110])
buf.toString('hex') // => "3830e3c6a7c200670098ff00000005000000073a6d6574686f6400000004504f5354000000083a76657273696f6e"
buf.toString() // => "80�Ƨ�g��:methodPOST:version"

var zlib = require('zlib')
zlib.inflate(buf) 
// => Error: Missing dictionary
//    at Zlib._handle.onerror (zlib.js:363:17)
```

**Comment by dignifiedquire** [comment]
Some more progress, inflating with the correct dictionary from `spdy-transports` results in a successful inflation. So it looks like the zlib in the browser is not using the correct dictionary.

**Comment by dignifiedquire** [comment]
So it turns out it's actually documented why this does not work. In the browser `zlib` is replaced with `browserify-zlib`, which [states in its readme](https://github.com/devongovett/browserify-zlib#not-implemented) that the `dictionary` option is not implemented. So node supports the dictionary and the browser doesn't, if browser <-> browser communication is tested they both ignore it and it works, but if you start sending data from node to the browser, the browser is unable to `inflate` resulting in this message as it's not able to handle the dictionary.

**Comment by diasdavid** [comment]
> 14:04 <dignifiedquire> daviddias: well this is the library you need to patch in the first place to get that done: https://github.com/nodeca/pako
>  14:05 <dignifiedquire> daviddias: https://github.com/nodeca/pako/issues/69#issuecomment-172870315 it was only due to time constraints they didn't add it, so my guess is they'd be happy to accept a PR

**Comment by indutny** [comment]
Looks like you figured it out, guys!

**Comment by indutny** [comment]
Why not use http2 in browser, though?

**Comment by diasdavid** [comment]
thank you anyways, @indutny :D 

We could and want to use http2 too, still have to create support for that in the go implementation

**Comment by dignifiedquire** [comment]
First implementation for pako is up here: https://github.com/nodeca/pako/pull/77 

cc @diasdavid

**Comment by diasdavid** [comment]
@dignifiedquire I saw this branch with the new browserify-zlib (https://github.com/ipfs/browserify-zlib/pull/1) running in your machine. Seems you never had the chance to push the commit. Could you do it now?

**Comment by dignifiedquire** [comment]
@diasdavid there where no changes, just installing the correct browserify-zlib does the trick

**Comment by diasdavid** [comment]
greeen finally :D

**Pull request merged**